src/server/twemcache-rs/src/request_processor.rs

```diff
@@ -0,0 +1,286 @@
+            MemcacheCommand::Get => self.process_get(request, write_buffer),
+            MemcacheCommand::Gets => self.process_gets(request, write_buffer),
+            MemcacheCommand::Set => self.process_set(request, write_buffer),
+            MemcacheCommand::Add => self.process_add(request, write_buffer),
+            MemcacheCommand::Replace => self.process_replace(request, write_buffer),
+            MemcacheCommand::Cas => self.process_cas(request, write_buffer),
+            MemcacheCommand::Delete => self.process_delete(request, write_buffer),
+        }
+    }
+
+    fn process_get(&mut self, request: MemcacheRequest, write_buffer: &mut BytesMut) {
+        let mut total = 0;
+        let mut found = 0;
+        increment_counter!(&Stat::Get);
+        for key in request.keys() {
+            if let Some(item) = self.data.get(key) {
+                let response = MemcacheResponse::Item { item, cas: false };
+                response.serialize(write_buffer);
+                found += 1;
+            }
+            total += 1;
+        }
+
+        increment_counter_by!(&Stat::GetKey, total);
+        increment_counter_by!(&Stat::GetKeyHit, found);
+        increment_counter_by!(&Stat::GetKeyMiss, total - found);
+
+        trace!(
+            "get request processed, {} out of {} keys found",
+            found,
+            total
+        );
+        MemcacheResponse::End.serialize(write_buffer);
+    }
+
+    fn process_gets(&mut self, request: MemcacheRequest, write_buffer: &mut BytesMut) {
+        let mut total = 0;
+        let mut found = 0;
+        increment_counter!(&Stat::Gets);
+        for key in request.keys() {
+            if let Some(item) = self.data.get(key) {
+                let response = MemcacheResponse::Item { item, cas: true };
+                response.serialize(write_buffer);
+                found += 1;
+            }
+            total += 1;
+        }
+
+        increment_counter_by!(&Stat::GetsKey, total);
+        increment_counter_by!(&Stat::GetsKeyHit, found);
+        increment_counter_by!(&Stat::GetsKeyMiss, total - found);
+
+        trace!(
+            "gets request processed, {} out of {} keys found",
+            found,
+            total
+        );
+        MemcacheResponse::End.serialize(write_buffer);
+    }
+
+    fn process_set(&mut self, request: MemcacheRequest, write_buffer: &mut BytesMut) {
+        increment_counter!(&Stat::Set);
+        let reply = !request.noreply();
+        let ttl = self.get_ttl(&request);
+        let key = request.keys().next().unwrap();
+        let value = request.value().unwrap_or(b"");
+        match self.data.insert(
+            key,
+            value,
+            Some(&request.flags().to_be_bytes()),
+            CoarseDuration::from_secs(ttl),
+        ) {
+            Ok(_) => {
+                increment_counter!(&Stat::SetStored);
+                if reply {
+                    MemcacheResponse::Stored.serialize(write_buffer);
```

Comment:
Related to my proposal in the worker(s) file, I recommend using a response object that's not serialized in this step. And make the serialization decision and carry it out one level higher.

Reply:
I've moved serialization up a level. Still working on the refactoring, but this specific recommendation is addressed